ystia/yorc-a4c-plugin
# Feature/brbdcf 973 services support

# Pull Request description

## Description of the change

Support of Alien services in Deployment topologies :
the Deployment topology can reference a node template of an abstract node type (external service), or of an non-existent node type (managed service) which is actually a reference to a service provided by an application whose provisioning and lifecycle is not managed in this deployment.
In this context, the use of an abstract or unknown node type should be accepted by the Orchestrator.

### What I did

Updated the Yorc plugin to add to node templates of class org.alien4cloud.tosca.model.templates.ServiceNodeTemplate, a directive "substitutable" as mentioned by the TOSCA specification, so that the orchestrator knows this node template references an external application, and so the associated node type is :
- either abstract (external service)
- or non-existent, but referencing a template name of an import providing a topology template of the Application providing the service.

Added a new topology modifier ServiceTopologyModifier, that will remove delegate operations in workflows on such node templates referencing a service and for which the Orchestrator won't manage the lifecycle.
Adding instead a state change step, to pass the state of this serive to Started, else the Runtime view would always show the corresponding node in the transient state Initial.

Added in tosca-samples, a Topology allowing to deploy a Consul Server Application as a managed service. Added a Client Application using this service, installing a consul agent that will join the Consul cluster connecting to the service provided by the Consul Server. 

### How to verify it

This Pull Request has to be used with the corresponding Yorc orchestrator implementing feature BRBDCF-973-services-support

Setup :

In Alien4Cloud, upload a Yorc plugin from this branch
Install a Yorc orchestrator implementing feature BRBDCF-973-services-support
Define a Hosts Pool location of at least two hosts, each having labels public_adress and private_address

From Catalog > Manage archives > Git import > Git Location
Create a Git location :
Repository URL: https://github.com/ystia/yorc-a4c-plugin
Branch: feature/BRBDCF-973-services-support
Archive folder: tosca-samples/org/ystia/yorc/samples/consul

Create an application using the topology template ConsulServiceTemplate. At the last step of Deployment, select the checkbox allowing to create a Service in Alien4Cloud for this deployment.
Once the deployment is done, click on the output variable web_ui_url and check in the Nodes tab that there is just one node in the cluster.

Create an application using the topology ConsulClientTemplate. Check the topology to verify this topology is using an abstract node type for the Consul Server.
Check the Matching step and verify this abstract node type was matched against the service created in the previous step. Deploy the application.
Once the deployment is done, go back to the previous Application deployment, click on the output variable web_ui_url and check in the Nodes tab that there are now two nodes in the cluster.
